### ISSUE TITLE
Use "okta.oauth2.issuer" consistently

### ISSUE THREAD

**Issue by fschmager** [okta/okta-spring-boot#33]
The property `okta.oauth2.issuer` is expected to be different in various parts of okta spring boot. 

During application startup the issuer must not end in `/oauth2/default` in order to be able to discover the OIDC endpoints. The discovery endpoint is `https://${okta.hostname}/.well-known/openid-configuration`.

This code is executed during startup:
`com.okta.spring.oauth.discovery.DiscoveryPropertySource#getDiscoveryMetadata()`
```java
String issuerUrl = environment.getRequiredProperty(OKTA_OAUTH_ISSUER);
OidcDiscoveryMetadata discoveryMetadata = createDiscoveryClient(issuerUrl).discover();
```
If `okta.oauth2.issuer` is configured to end in `/oauth2/default` (as documented) the following  stacktrace is the result. This stems from the fact that the discovery endpoint is _not_ available under `https://${okta.hostname}/oauth2/default/.well-known/openid-configuration`
```
Caused by: org.springframework.web.client.HttpClientErrorException: 404 Not Found
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:85) ~[spring-web-4.3.13.RELEASE.jar:4.3.13.RELEASE]
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:707) ~[spring-web-4.3.13.RELEASE.jar:4.3.13.RELEASE]
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:660) ~[spring-web-4.3.13.RELEASE.jar:4.3.13.RELEASE]
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:635) ~[spring-web-4.3.13.RELEASE.jar:4.3.13.RELEASE]
	at org.springframework.web.client.RestTemplate.getForObject(RestTemplate.java:310) ~[spring-web-4.3.13.RELEASE.jar:4.3.13.RELEASE]
	at com.okta.spring.oauth.discovery.OidcDiscoveryClient.discover(OidcDiscoveryClient.java:49) ~[okta-spring-security-oauth2-0.3.0.jar:0.3.0]
	at com.okta.spring.oauth.discovery.DiscoveryPropertySource.getDiscoveryMetadata(DiscoveryPropertySource.java:105) ~[okta-spring-security-oauth2-0.3.0.jar:0.3.0]
	at com.okta.spring.oauth.discovery.DiscoveryPropertySource.getProperty(DiscoveryPropertySource.java:77) ~[okta-spring-security-oauth2-0.3.0.jar:0.3.0]
	at org.springframework.core.env.PropertySourcesPropertyResolver.getProperty(PropertySourcesPropertyResolver.java:81) ~[spring-core-4.3.13.RELEASE.jar:4.3.13.RELEASE]
	at org.springframework.core.env.PropertySourcesPropertyResolver.getProperty(PropertySourcesPropertyResolver.java:66) ~[spring-core-4.3.13.RELEASE.jar:4.3.13.RELEASE]
	at org.springframework.core.env.AbstractEnvironment.getProperty(AbstractEnvironment.java:537) ~[spring-core-4.3.13.RELEASE.jar:4.3.13.RELEASE]
	at org.springframework.boot.bind.RelaxedPropertyResolver.getProperty(RelaxedPropertyResolver.java:84) ~[spring-boot-1.5.9.RELEASE.jar:1.5.9.RELEASE]
	at org.springframework.boot.bind.RelaxedPropertyResolver.getProperty(RelaxedPropertyResolver.java:64) ~[spring-boot-1.5.9.RELEASE.jar:1.5.9.RELEASE]
	at org.springframework.boot.autoconfigure.security.oauth2.resource.ResourceServerTokenServicesConfiguration$JwkCondition.getMatchOutcome(ResourceServerTokenServicesConfiguration.java:379) ~[spring-boot-autoconfigure-1.5.9.RELEASE.jar:1.5.9.RELEASE]
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:47) ~[spring-boot-autoconfigure-1.5.9.RELEASE.jar:1.5.9.RELEASE]
	... 26 common frames omitted
```
So, in order to get an app to successfully startup `okta.oauth2.issuer` must not end in `/oauth2/default`. 

However, the code validating access token expects the issuer to end in `/oauth2` (also not as documented) to construct a valid keys URL.

The keys URL is constructed during configuration:
`com.okta.spring.oauth.OktaTokenServicesConfig.LocalTokenValidationConfig#tokenStore()`:
```java
return new JwkTokenStore(oktaOAuth2Properties.getIssuer() + "/v1/keys", accessTokenConverter(), jwtClaimsSetVerifier());
```
For this to work the issuer must be `https://${okta.hostname}/oauth2` (no trailing `default` as documented).
If `/oauth2` is omitted (to get a successful startup and discovery) the access token validation will fail:
```
2018-01-11 17:15:07.485 DEBUG 82456 --- [nio-8080-exec-1] p.a.OAuth2AuthenticationProcessingFilter : Authentication request failed: error="server_error", error_description="An I/O error occurred while reading from the JWK Set source: https://dev-xxxxxx.oktapreview.com/v1/keys"
2018-01-11 17:15:07.533 DEBUG 82456 --- [nio-8080-exec-1] s.s.o.p.e.DefaultOAuth2ExceptionRenderer : Written [error="server_error", error_description="An I/O error occurred while reading from the JWK Set source: https://dev-xxxxxx.oktapreview.com/v1/keys"] as "application/json;charset=UTF-8" using [org.springframework.http.converter.json.MappingJackson2HttpMessageConverter@287e7668]
```

I'm wondering if `OktaTokenServicesConfig` should be using the keys url from the discovery step instead of manually constructing it. Otherwise, what's the point of doing the discovery if the discovered URLs are not used.

okta-spring-boot.version: 0.3.0
spring-security-oauth2.version: 2.2.1.RELEASE

`application.yml`
```yml
okta:
  tenant: TTT
  hostname: ${okta.tenant}.oktapreview.com
okta.oauth2:
  #issuer: https://${okta.hostname} # successful discovery but failing access token validation
  #issuer: https://${okta.hostname}/oauth2 # failing OIDC discovery, needed for access token validation
  #issuer: https://${okta.hostname}/oauth2/default # as documented but not as the code is working
  clientId: XXX
```

**Comment by fschmager** [comment]
Nevermind. I missed the step about creating an Authorization Server. I used the app's issuer instead of the auth server's.
Maybe there is room for improving the documentation?

**Comment by fschmager** [comment]
I followed the [README.md](https://github.com/okta/okta-spring-boot/blob/master/README.md) which told me 1) to have a dev account. 
I've created a developer account quite a while ago (shortly before Stormpath had been acquired). Maybe that had something to do with me not having a "default" Auth Server. Once I created one and put the Auth Server ID into the issuer things started to work.

Maybe it should be pointed out that people ensure they really do have an Auth Server. Default, custom or otherwise.
